n8n-io/n8n
# fix(core): Reduce large relay message payloads to protect Redis

## Summary

It is possible for a manual execution in scaling mode to cause so much data to flow through Redis that it becomes unresponsive. Specifically, the `nodeExecuteAfter` and `executionFinished` events may carry arbitrarily large payloads. 

In this PR we replace payloads above a 5 MB limit with placeholders that signal the UI to direct the user to view the full payload in execution history instead. Events with payloads below the size limit continue to be sent and displayed as usual.

In future, we can override these placeholders with the full execution data fetched from the DB at the end of the execution. Since we are already planning to refactor execution data handling on the client in the near future, that change is out of scope for this PR.

### Testing

- Start a main in scaling mode with `OFFLOAD_MANUAL_EXECUTIONS_TO_WORKERS=true`
- Start a worker
- Monitor Redis: `docker exec -it <container> redis-cli monitor`
- Start an ad hoc server (see Slack) at `localhost:3000` serving a 30 MB payload
- Import workflow (see Slack), run it manually and call the test webhook
- Verify that worker is processing manual execution and relaying to main
- Verify that items after the `Fetch large payload` node have been trimmed
- Verify in execution history that those items are displayed in full

Recording: https://share.cleanshot.com/GCrCsPXw


## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/PAY-2403/reduce-large-relay-message-payloads-to-protect-redis

## Review / Merge checklist

- [ ] ~~PR labeled with `release/backport`~~
